### ISSUE TITLE
Unexpected clones

### ISSUE THREAD

**Issue by xethorn** [elastic/elasticsearch-dsl-py#138]
First, thanks for this awesome library. 

While I was playing with it today, it didn't look obvious (or maybe I've missed it from the documentation) how non chaining works – and the fact that clones of the object are always created whenever you filter or query.

For instance:

```python
current_search = Search(using=elastic, index="my-index") \
    .filter("term", category="search") \
    .query("match", title="python")   \
    .query(~Q("match", description="beta"))

current_search.aggs.bucket('per_tag', 'terms', field='tags') \
    .metric('max_lines', 'max', field='lines')
```

Generates:
```
{'aggs': {'per_tag': {'aggs': {'max_lines': {'max': {'field': 'lines'}}},
   'terms': {'field': 'tags'}}},
 'query': {'filtered': {'query': {'bool': {'must_not': [{'match': {'description': 'beta'}}],
     'must': [{'match': {'title': 'python'}}]}},
   'filter': {'term': {'category': 'search'}}}}}
```

I assumed that writing the following code would be identical (because the way it's written – you perform an action an object.)

```python
s = Search(using=elastic, index="my-index")

if True:
    s.filter("term", category="search")

s.query("match", title="python")   \
    .query(~Q("match", description="beta")) \
    .aggs.bucket('per_tag', 'terms', field='tags') \
    .metric('max_lines', 'max', field='lines')
```

Generates:
```
{'query': {'match_all': {}}}
```

Looking at the code, it seems like the object clones itself (and not necessary consistently, aggregations).

```python
s = Search(using=elastic, index="my-index")

if True:
    s = s.filter("term", category="search")

s = s.query("match", title="python")   \
    .query(~Q("match", description="beta"))

s.aggs.bucket('per_tag', 'terms', field='tags') \
    .metric('max_lines', 'max', field='lines')
```

Returns:

```
{'aggs': {'per_tag': {'aggs': {'max_lines': {'max': {'field': 'lines'}}},
   'terms': {'field': 'tags'}}},
 'query': {'filtered': {'query': {'bool': {'must_not': [{'match': {'description': 'beta'}}],
     'must': [{'match': {'title': 'python'}}]}},
   'filter': {'term': {'category': 'search'}}}}}
```

Basically, how do you recommend writing code that has a lot of if / else statements and potential calls to external methods?

**Comment by HonzaKral** [comment]
I do recommend doing things like:

    if apply_filter:
        s = s.filter(...)

If you want to pass the Search into another function/method and you expect the method to alter it, it should be a return value: `s = apply_filters(s)`.

The aggregations exception is a compromise between this concept and some convenience when constructing more complex aggregations. I wasn't able to come up with a better solution there, though I am open to suggestions.

This also means that you cannot mix queries/filters with aggregations on the same line.

It is a compromise and I would welcome any pointers on how to make it better or explain it better.

Thanks for raising this issue.

**Comment by xethorn** [comment]
I'm definitely missing a lot of business knowledge (only started to use the class a few hours ago.) If you want to chain, this can easy be done in python while returning the same object (see below) – so I believe there might be a case where it's not possible - do you have a code sample I could take a shot at?

Until then, it might be very useful for people who are consuming the library to have a part that explicitly mentions this particular behavior. What do you think?

```python
class Search:
    def __init__(self, *args):
        self.args = args

    def add(self, *args):
        self.args = self.args + args
        return self

    def multiple(self, number):
        self.args = self.args * number
        return self

s = Search('a', 'b').add('c', 'd').multiple(2)
print(s.args)

s = Search('a', 'b')
s.add('c', 'd')
s.multiple(2)
print(s.args)
```

**Comment by HonzaKral** [comment]
We went with the cloning for several reasons:

* it makes it safe when handing your search objects to third party code (explicitly returning the muted value is required, explicit is better than implicit ;) )
* it's a known pattern (for example django querysets work the same way)
* it allows for easy branching - start with a global base search (including for example the doc_types to be searched, global filters - like filtering out unpublished articles - etc) and keep adding queries/filters to it before every execution

I still feel that it's worth it, both the extra code and the slight interface complication for users, to avoid weird errors like mutating a global object. The unfortunate thing is that we cannot be consistent with `aggs` because we wouldn't be able to easily define nested aggregations otherwise.

We came up with this compromise after a lot of discussions. I am not 100% happy with it, but I am not unhappy with it either and we haven't seen many issues with it - most of it is solvable by better documentation.

What do you think?

**Comment by xethorn** [comment]
It's a pretty interesting problem to have. The current implementation is always duplicating the Search object. It means that the query from the example is creating a minimum of 3 clones (which adds latency.) 

From an engineering perspective:

```python
s = Search(using=elastic, index="my-index")

if True:
    s.filter("term", category="search")

s.query("match", title="python")   \
    .query(~Q("match", description="beta")) \
    .aggs.bucket('per_tag', 'terms', field='tags') \
    .metric('max_lines', 'max', field='lines')
```

Should be equal to:

```python
current_search = Search(using=elastic, index="my-index") \
    .filter("term", category="search") \
    .query("match", title="python")   \
    .query(~Q("match", description="beta"))

current_search.aggs.bucket('per_tag', 'terms', field='tags') \
    .metric('max_lines', 'max', field='lines')
```

To go back to the initial points:

_it makes it safe when handing your search objects to third party code (explicitly returning the muted value is required, explicit is better than implicit ;) )_

:+1:  That said: when does this need to happen? It seems like the Search object should only become immutable after the search has been executed. So in this case, if you try to create a mutation of the initial object, you should probably raise an exception that suggests to use `.clone`.

_it allows for easy branching - start with a global base search (including for example the doc_types to be searched, global filters - like filtering out unpublished articles - etc) and keep adding queries/filters to it before every execution_

It's definitely a good idea to keep around. To go back to the example:

```python
s = Search(using=elastic, index="my-index")

if True:
    s.filter("term", category="search")

s.query("match", title="python")   \
    .query(~Q("match", description="beta")) \
    .aggs.bucket('per_tag', 'terms', field='tags') \
    .metric('max_lines', 'max', field='lines')

s.execute() # makes s immutable

# raise exception
s.filter('term', title='something') 

 # Create a clone and set a new filter on the clone.
new_search = s.clone.filter('term', title='something')
```

Another route could be to have this (so you can share aggregation configurations and more):

```python
Search(
    Filter('term', category='search'),
    Query('match', title='python'),
    Query(~Q('match', description='beta'))
    Aggregation(
        'per_tag',
        'terms', 
        field='tags',
        aggregations=[
            Metric('max_lines', 'max', field='lines')])
```

I'm not a big fan (it feels similar to writing the json itself, and you will need to make sure that all your objects are immutable, similar to what is done today in the elasticsearch-dsl) but you could do this at the end: create a list, and provide the list:

```python
args = [Filter(), Query(), Query(), Aggregation()]
search = Search(*args)
search.execute()
```


What do you think?

**Comment by HonzaKral** [comment]
The constructor solution won't help much, the problem is not with creating the queries, it's adding more queries/filters etc to an existing search object. That's what's hard when working with the raw dict-based dsl and what I wanted to abstract away. You can also do this now:

    s = Search()
    s.query = Q('match') + ~Q('match')
    s.filter = F('term')
    s.aggs.bucket('per_tag', A('terms', field='tags', aggs={'max_lines': A('max', field='lines')}))

That will do everything that you wish without any cloning of anything under the hood. If you are concerned about the cloning overhead, this is definitely an option.

Other than that I have nothing to add to your points - you are right. It's just a matter of opinion and style and so far everybody I talked to (that includes talking to myself) prefers the cloning mechanism to mutate-in-place + add explicit .clone() method. Even the, admittedly half-baked, version we have. When I did the first version without the cloning many people were surprised that `s.query().filter().query()` altered the original `Search` object.

I did some profiling when making the decision - the `_clone` method takes the same amount of time regardless of the complexity of the queries/filters/... and only grows with extra parameters like pagination, extra, fields and some others. The average time for a run is slightly above 30 micro seconds. Even if that were to occur 10 times before a query is executed it would still add an overhead of just a third of a millisecond. Compared to the network trip that will follow (not to mention the actual query execution time) I would consider that negligible.

I find the user expectations, least surprise and documentation to be a much more important issue. Especially since the people that care about microseconds have tools at their disposal to make it faster.

Does that make sense?

Thanks for the interesting discussion, I am always super happy for great feedback like this!

**Comment by xethorn** [comment]
Quick reply!

The last code sample you've sent is definitely good (that's very close to what I was looking for). Re: documentation: you're right, documenting the different approaches (which would set better expectations on how the framework works) will be definitely useful.

This conversation has definitely been very good, thanks for your time! :+1:
